### PR TITLE
#56 Create Word entities from the word template

### LINKED ISSUE #56: Create Words in Session from .json template word lists

### PR DESCRIPTION
**Pull request by catfinch**
Read the selected word list file and create the Words in the background

resolves #56

### CODE REVIEW

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 1]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+        protected List<Long> doInBackground(Session... sessions) {
+            return sDAO.insertSession(sessions);
+        }
+
+    }
+
+    /**
+     * Async task for inserting words from the selected word list
+     */
+    private static class InsertWordsTask extends AsyncTask<Void, Double, List<Long>> {
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: This task is logically involved as part of the process of creating a session.  I would recommend merging this into `InsertSessionsTask` to reduce the complexity of chaining code.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 2]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+        }
+
+        @Override
+        protected List<Long> doInBackground(Void... voids) {
+            List<Long> ids = new ArrayList<>();
+            try {
+                // Wait for the insert session task to finish and get the result (the id(s) of the session)
+                ids = task.get();
+            } catch (InterruptedException | ExecutionException e) {
+                Log.getStackTraceString(e);
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: This is a noop.  Please at least send the message to the log.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 3]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+            if (ids.size() == 1) {
+                sessionID = ids.get(0);
+            }
+            insertFromAsset();
+
             return null;
         }
+
+        @Override
+        protected void onProgressUpdate(Double... values) {
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: How long is this expected to take?  Are we talking a second?  Ten's of seconds?  Minutes?

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 4]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+
+                    // Insert
+                    wordDao.insertWords(word);
+                    this.publishProgress((double) i);
+                }
+
+                // Mark to commit the changes to the DB
+                sessionActivity.get().db.setTransactionSuccessful();
+            } catch (Exception e) {
+                Log.getStackTraceString(e);
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Noop.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 5]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+        private String loadWordList() {
+            String file = "";
+            try {
+                InputStream is = sessionActivity.get().getApplicationContext().getAssets().open("wordLists/" + sessionActivity.get().worldListToLoad);
+                byte[] buffer = new byte[is.available()];
+                is.read(buffer);
+                is.close();
+                file = new String(buffer, "UTF-8");
+            } catch (Exception e) {
+                Log.getStackTraceString(e);
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Noop.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 6]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+
+    }
+
+    /**
+     * Async task for inserting words from the selected word list
+     */
+    private static class InsertWordsTask extends AsyncTask<Void, Double, List<Long>> {
+        private WordDao wordDao;
+        private AsyncTask<Session, Void, List<Long>> task;
+        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: What is your logic behind using a `ThreadLocal`?  The activity does not belong to this thread.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 7]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+            CharSequence sequence = new DecimalFormat("##.##%").format(progress);
+            // Too slow!!! TODO get quick progress updates
+//            Toast.makeText(sessionActivity.getApplicationContext(), sequence, Toast.LENGTH_SHORT).show();
+        }
+
+        /**
+         * Insert the words from the selected asset or exit if none selected
+         */
+        private void insertFromAsset() {
+            if (sessionActivity.get().worldListToLoad.equals("")) {
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Just `.get()` it once per function and store it in a variable.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 8]
```diff
@@ -118,18 +130,121 @@ public void save_settings_fab_pressed(View view) {
...
+        /**
+         * Read the selected file
+         *
+         * @return The contents of the file
+         */
+        private String loadWordList() {
+            String file = "";
+            try {
+                InputStream is = sessionActivity.get().getApplicationContext().getAssets().open("wordLists/" + sessionActivity.get().worldListToLoad);
+                byte[] buffer = new byte[is.available()];
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: This is not guaranteed to be the entire file.  Probably safer to loop with a `StringBuilder`.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java (hunk 6 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Lint was warning about a data leak. It was something to do with using part of an activity in a static class

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java (hunk 3 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Depends on the size of the word list. The smallest sample one given to use causes a noticeable hang if we wait. I put that there for when we have some way of showing progress for longer files

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java (hunk 7 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: How about getting it only when I need the context (which has the data leak), and the other variables are stored on construction

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java (hunk 6 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: When the tasks got merged, I keep getting `null` assigned to the variable. I'm going to ignore the warning and put a TODO tag there

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 9]
```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
...
+    private static class InsertSessionsTask extends AsyncTask<Session, Double, Void> {
         private SessionDao sDAO;
+        private WordDao wordDao;
+        // TODO: fix null object reference so we don't risk a data leak with the context
+//        private final ThreadLocal<SessionActivity> sessionActivity = new ThreadLocal<>();
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
+        private final Context context;
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: I see.  It is `WeakReference` you want, not `final` (you could use that if you want access to the entire activity as well).

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 10]
```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
...
+            wordDao = db.wordDao();
             sDAO = db.sessionDao();
         }
 
         @Override
         protected Void doInBackground(Session... sessions) {
-            sDAO.insertSession(sessions);
+            List<Long> ids = sDAO.insertSession(sessions);
+
+            if (ids.get(0) != null) {
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: This will crash if no sessions where inserted.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 11]
```diff
@@ -118,18 +129,127 @@ public void save_settings_fab_pressed(View view) {
...
+                insertFromAsset();
+            }
             return null;
         }
+
+        /**
+         * Update the main thread (the UI) with some indication of progress
+         */
+        @Override
+        protected void onProgressUpdate(Double... values) {
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Since it sounds you estimate this to be less than a second, I do not believe creating a progress bar will be necessary.  The word lists are unlikely to be much larger than what we have been supplied with since they are expected to be able to be completed during a single session.

I would recommend just dropping this function.

File: app/src/main/java/org/sil/gatherwords/room/Session.java [hunk 12]
```diff
@@ -17,7 +17,7 @@
 public class Session {
 	// Define the primary key
 	@PrimaryKey(autoGenerate = true)
-	public int id = 0;
+	public Long id;
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/room/Session.java]: What is the purpose of this change?  If you want a larger key space, why not use the primitive `long`?  Does this schema change require changing the db version?

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java (hunk 10 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: I'll fix it, but when would this line run if there are no functions. The only entry point creates the session being inserted

File: app/src/main/java/org/sil/gatherwords/room/Session.java (hunk 12 above)
Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/room/Session.java]: `int` wasn't an option for returning the id's after an insert

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 13]
```diff
@@ -118,18 +128,110 @@ public void save_settings_fab_pressed(View view) {
...
+        private WeakReference<SessionActivity> sessionActivityReference;
+        private Long sessionID;
+        private int maxProgress;
+        private AppDatabase db;
+        private String wordList;
 
-        InsertSessionsTask(AppDatabase db) {
+        InsertSessionsTask(SessionActivity activity) {
+            sessionActivityReference = new WeakReference<>(activity);
+            db = AppDatabase.get(sessionActivityReference.get().getApplicationContext());
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: No need for indirection here; we can just `db = AppDatabase.get(activity.getApplicationContext());`.

File: app/src/main/java/org/sil/gatherwords/SessionActivity.java [hunk 14]
```diff
@@ -118,18 +128,110 @@ public void save_settings_fab_pressed(View view) {
...
+         * Read the selected file
+         *
+         * @return The contents of the file
+         */
+        private String loadWordList() {
+            InputStream is = null;
+            BufferedReader br = null;
+            String file = null;
+            try {
+                is = sessionActivityReference.get().getApplicationContext().getAssets().open("wordLists/" + wordList);
```
Comment (alpha0010) [inline app/src/main/java/org/sil/gatherwords/SessionActivity.java]: `WeakReference` may be garbage collected at any time, so this could throw a npe.

Comment (catfinch) [reply app/src/main/java/org/sil/gatherwords/SessionActivity.java]: Then what's the point of using the `WeakReference` type if I can't use it to avoid the possible data leak from storing a `static Context` variable?  
What other options will work?

**Comment by alpha0010** [comment]
Merged in cd30d4198795ed5e4caa19ac59ca1d272ec5c561.
